srml/offences/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Full identification of the validator.
+	type IdentificationTuple: Parameter;
+	/// A handler called for every offence report.
+	type OnOffenceHandler: OnOffenceHandler<Self::AccountId, Self::IdentificationTuple>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Offences {
+		/// A mapping between unique `TimeSlots` within a particular session and the offence `Kind` into
+		/// a vector of offending authorities.
+		///
+		/// It's important that we store all authorities reported for an offence for any kind and
+		/// timeslot since the slashing will increase based on the length of this vec.
+		OffenceReports get(offence_reports):
+			double_map Kind, blake2_256((SessionIndex, TimeSlot))
+			=> Vec<OffenceDetails<T::AccountId, T::IdentificationTuple>>;
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		/// There is an offence reported of the given `kind` happened at the `session_index` and
+		/// (kind-specific) time slot. This event is not deposited for duplicate slashes.
+		Offence(Kind, SessionIndex, TimeSlot),
+	}
+);
+
+decl_module! {
+	/// Offences module, currently just responsible for taking offence reports.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+	}
+}
+
+impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T::IdentificationTuple, O>
+	for Module<T> where
+	T::IdentificationTuple: Clone,
+{
+	fn report_offence(reporter: Option<T::AccountId>, offence: O) {
+		let offenders = offence.offenders();
+		let time_slot = offence.time_slot();
+		let session = offence.session_index();
+		let validator_set_count = offence.validator_set_count();
+
+		// Check if an offence is already reported for the offender authorities
+		// and otherwise stores that report.
+		let mut changed = false;
+		let all_offenders = <OffenceReports<T>>
+			::mutate(&O::ID, &(session, time_slot), |offending_authorities| {
+				for offender in offenders {
+					// TODO [slashing] This prevents slashing for multiple reports of the same kind at the same slot,
+					// note however that we might do that in the future if the reports are not exactly the same (dups).
+					// De-duplication of reports is tricky though, we need a canonical form of the report
+					// (for instance babe equivocation can have headers swapped).
+					if !offending_authorities.iter().any(|details| details.offender == offender) {
+						changed = true;
+						offending_authorities.push(OffenceDetails {
+							offender,
+							count: 0,
+							reporters: reporter.clone().into_iter().collect(),
+						});
+					}
+				}
+
+				if changed {
+					for details in offending_authorities.iter_mut() {
+						details.count += 1;
+					}
+				}
+
+				offending_authorities.clone()
+			});
+
+		if !changed {
+			// The report contained only duplicates, so there is no need to slash again.
+			return
```

Comment:
shouldn't fully-duplicate slashing reports return an error of some kind, so that the transaction-processing later on can reject the transaction?

in the event of an equivocation, we should have many many reports floating around the network. it would be useless to include all of them in blocks. We might want to include the first few to lower variance a bit.

there might be some possible use of the `TaggedTransactionQueue` to ensure that inclusion of one immediately invalidates others.

Reply:
Yes, we should definitely use `TaggedTransactionQueue` to deduplicate `UnsignedTransaction`s containing the same reports. (afair we already do that for `imonline` reports).
Afaict, with babe/grandpa we will use regular `SignedTransactions` to report equivocation, so the callers need to pay for the execution anyway. As future optimization we should probably change the reporting code to avoid generating the report if we find that there is already some other report for this offence in the pool.

Reply:
I don't think we'll be able to dis-incentivize front-running. The way we've structured things now, there's going to be a mad rush every time there's a misbehavior and variance over whether you actually get paid out over a report will be really high.

There's no point in including such wasted transactions in blocks, though.